Table name ids

doowruc opened the pull request:
Support for db schema where id columns include the name of the table. 

e.g.: 

```
Table: Parent
Columns: ParentId, Name

Table: Child
Columns: ChildId, ParentId, Name
```

doowruc committed:
Added TableNameIds topic branch

doowruc committed:
Implement support for relationships when table Id columns are named after the table

- In IdPropertySetterInterceptor: If getting property "Id" fails, try to get propertyName + "Id"

- In ParentPropertySetterInterceptor: If getting parent property "Id" fails, try to get parentProperty.Name + "Id"

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/CachedEntityTypesProviderTest.cs:
```diff
@@ -1,46 +1,71 @@
+            var result = sut.GetTypes();
+
+            result.Should().BeEquivalentTo(types);
+        }
```
Is this test necessary? It appears that it mimics `SutIsDecorator` (above)

doowruc replied:
I have added the 3 new classes to the original `SutIsDecorator` and removed this

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/CachedEntityTypesProviderTest.cs:
```diff
@@ -1,46 +1,71 @@
+            resultSecond.Should().BeEquivalentTo(typesFirst)
+                .And.BeEquivalentTo(resultFirst);
+        }
+    }
```
Also appears to be a duplicate (see `Types EnumeratedOnlyOnce`)

doowruc replied:
I have added the 3 new classes to the original `TypesEnumeratedOnlyOnce` and removed this

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+        public void IdShouldMatchNavigationPropertyIdWhenTableNameId(Far far)
+        {
+            far.BooId.Should().Be(far.Boo.BooId);
+        }
```
This one is valid

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+        public void ParentShouldBeSameObjectWhenTableNameId(Far far)
+        {
+            far.Boo.Far.Should().BeSameAs(far);
+        }
```
I don't think this tests the new functionality

doowruc replied:
agreed, removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+        {
+            boo.Qixes.Clear();
+            boo.Qixes.Should().HaveCount(0);
+        }
```
I don't think this tests the new functionality

doowruc replied:
agreed, removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+        public void GeneratedCollectionIsNotEmptyWhenTableNameId(Boo boo)
+        {
+            boo.Qixes.Should().NotBeEmpty();
+        }
```
I don't think this tests the new functionality

doowruc replied:
agreed, removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+        public void GeneratedPropertyIsNotNullWhenTableNameId(Far far)
+        {
+            far.Boo.Should().NotBeNull();
+        }
```
I don't think this tests the new functionality

doowruc replied:
agreed, removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+        public void FixtureCanCreateEFEntityWhenTableNameId(Far far)
+        {
+            far.Should().NotBeNull();
+        }
```
I don't think this tests the new functionality

doowruc replied:
agreed, removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+        public void ParentIdShouldBeSameWhenTableNameId(Far far)
+        {
+            far.Boo.FarId.Should().Be(far.FarId);
+        }
```
This one also looks valid

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+            foreach (var qix in boo.Qixes)
+            {
+                (qix.Boo == boo).Should().BeTrue();
+            }
```
I think this should test that `qix.BooId.Should().Be(boo.BooId)`

doowruc replied:
Updated

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+            var boo2 = far.Boo;
+
+            boo1.Should().BeSameAs(boo2);
+        }
```
I don't think this tests the new functionality

doowruc replied:
agreed, removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+        public void UnfrozenTypeYieldsDifferentObjectWhenTableNameId(Boo boo, Far far)
+        {
+            far.Boo.Should().NotBeSameAs(boo);
+        }
```
I don't think this tests the new functionality

doowruc replied:
agreed, removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/IntegrationTest.cs:
```diff
@@ -88,6 +88,79 @@ public void UnfrozenTypeYieldsDifferentObject(Bar bar, Foo foo)
+        [Theory, AutoEFData(typeof(MockDbContext))]
+        public void FrozenTypeYieldsSameObjectWhenTableNameId([Frozen] Boo boo, Far far)
+        {
+            far.Boo.Should().BeSameAs(boo);
```
I don't think this tests the new functionality

doowruc replied:
agreed, removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/Interception/IdPropertySetterInterceptorTest.cs:
```diff
@@ -38,6 +38,21 @@ public void SetsIdOnMatchingProperty(SUT sut, IInvocation invocation, int barId)
+
+            sut.Intercept(invocation);
+
+            boo.BooId.Should().Be(booId);
```
:+1:

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/Interception/ParentPropertySetterInterceptorTest.cs:
```diff
@@ -51,6 +51,34 @@ public void SetsParentIdWhenPropertyExists(SUT sut, IInvocation invocation, int
+            sut.Intercept(invocation);
+
+            boo.Far.Should().BeSameAs(far);
+        }
```
I don't think this tests the new functionality

doowruc replied:
agreed, removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/Interception/ParentPropertySetterInterceptorTest.cs:
```diff
@@ -51,6 +51,34 @@ public void SetsParentIdWhenPropertyExists(SUT sut, IInvocation invocation, int
+
+            sut.Intercept(invocation);
+
+            boo.FarId.Should().Be(far.FarId);
```
:+1:

alexfoxgill commented on src/AutoFixture.AutoEF/Interception/IdPropertySetterInterceptor.cs:
```diff
@@ -22,6 +22,11 @@ public void Intercept(IInvocation invocation)
+            {
+                proxyIdProp = invocation.ReturnValue.GetType().GetProperty(propertyName + "Id", bindingFlags);
+            }
+
```
:+1:

alexfoxgill commented on src/AutoFixture.AutoEF/Interception/ParentPropertySetterInterceptor.cs:
```diff
@@ -45,6 +45,11 @@ where prop.PropertyType.IsInstanceOfType(parent)
+            {
+                parentIdProperty = parent.GetType().GetProperty(parentProperty.Name + "Id");
+            }
+
```
:+1:

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/NavigationPropertyRequestSpecificationTest.cs:
```diff
@@ -67,6 +67,50 @@ public void ReturnsTrueWhenPropertyIsCollection([Frozen] IEntityTypesProvider ty
+
+            var result = sut.IsSatisfiedBy(property);
+
+            result.Should().BeTrue();
```
I'm not sure these tests need to exist - it looks like the only changed functionality is in `IdPropertySetterInterceptor` and `ParentPropertySetterInterceptor`, in which case the `NavigationPropertyRequestSpecification` will not have been touched

doowruc replied:
removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/Interception/SetPropertyReturnValueInterceptorTest.cs:
```diff
@@ -42,6 +42,24 @@ public void SetsPropertyOnValueReturn(SUT sut, IInvocation invocation)
+            sut.Intercept(invocation);
+
+            // Verify outcome
+            far.Boo.Should().Be(boo);
```
I'm not sure these tests need to exist - it looks like the only changed functionality is in `IdPropertySetterInterceptor` and `ParentPropertySetterInterceptor`, in which case the `SetPropertyReturnValueInterceptor` will not have been touched

doowruc replied:
removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/Interception/PropertyGetterInterceptionPolicyTest.cs:
```diff
@@ -44,6 +44,27 @@ public void ReturnsFalseOnSetter(SUT sut, IInvocation invocation)
+
+            var result = sut.ShouldIntercept(invocation);
+
+            result.Should().BeFalse();
```
I'm not sure these tests need to exist - it looks like the only changed functionality is in `IdPropertySetterInterceptor` and `ParentPropertySetterInterceptor`, in which case the `PropertyGetterInterceptionPolic` will not have been touched

doowruc replied:
removed

alexfoxgill commented:
Hi @doowruc, thanks for creating this pull request and taking the time to write all of the tests! As you can see from my comments, I suspect that some of them are unnecessary duplications of other tests, which can cloud the codebase. I'd be more than happy to merge your PR if you can either remove those tests, or point out what I'm missing!

doowruc commented:
I wanted to make sure that my addition didn't break existing functionality and I'm not familiar with the test framework you are using so I wasn't entirely sure how the tests are working! 

So, I simply replicated every test that had a Foo with a Far test, Bar with Boo and Qux with Qix.

In fact only 4 of the new tests failed before I added my changes (TDD FTW!), the rest passed straight away.

I'll remove the ones you have pointed out and make another commit.

doowruc committed:
Removed / updated tests as per code review

doowruc commented:
All done!

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/Interception/PropertyGetterInterceptionPolicyTest.cs:
```diff
@@ -1,49 +1,60 @@
+
+            result.Should().BeTrue();
+        }
+    }
```
I believe this is still a duplication of `ReturnsTrueOnGetter`

doowruc replied:
Removed

alexfoxgill commented on src/AutoFixture.AutoEF.Tests/NavigationPropertyRequestSpecificationTest.cs:
```diff
@@ -1,72 +1,105 @@
+
+            result.Should().BeTrue();
+        }
+    }
```
I believe this file should remain unchanged from the original, as the tests only concern `NavigationPropertyRequestSpecification` which was not modified

doowruc replied:
reverted

doowruc committed:
Removed another test as per code review

alexfoxgill commented:
Thanks for that although I think you missed one or two - I have pointed them out above.

Just one more thing - some of the original tests are appearing as changed in the diff when they're actually unmodified. My guess is that the line endings are different. If you're comfortable reverting those changes (leaving only the added tests) I'd be grateful. If not then it's no huge deal!

doowruc committed:
Further code review - removed more tests, corrected line endings back to Unix style

doowruc commented:
All done. I've removed those last couple of extra tests and corrected all the line endings (yay Notepad++)

Sorry to have caused so much hassle over a quite small change!

alexfoxgill commented:
Wonderful, thank you very much - I'll update the NuGet package today.

doowruc commented:
Does the Conventions section of the ReadMe need updating? Do you want me to do it?

alexfoxgill commented:
Good point, I've just updated it

alexfoxgill commented:
Sorry for the delay, the NuGet package is now updated to version 0.4.0 which includes this change. Thank you @doowruc for your contribution!
